src/modules/displaymanager/main.py

```diff
@@ -633,41 +636,57 @@ def desktop_environment_setup(self, default_desktop_environment):
-        greeter_path = os.path.join(
-            self.root_mount_point, "usr/share/xgreeters"
-            )
+        Returns a path *into the mounted target* of the preferred greeter -- usually
+        a .desktop file that specifies where the actual executable is. May return
+        None to indicate nothing-was-found.
+        """
+        greeters_dir = "usr/share/xgreeters"
+        greeters_target_path = os.path.join(self.root_mount_point, greeters_dir)
+        available_names = os.listdir(greeters_target_path)
+        available_names.sort()
+        desktop_names = [n for n in self.preferred_greeters if n in available_names] # Preferred ones
+        if desktop_names:
+            return desktop_names[0]
+        desktop_names = [n for n in available_names if n.endswith(".desktop")] # .. otherwise any .desktop
+        if desktop_names:
+            return desktop_names[0]
+        desktop_names = [n for n in available_names if not n.startswith(".")] # .. otherwise any non-dot-file
+        if desktop_names:
+            return desktop_names[0]
+        if available_names: # Anything?
+            return available_names[0]
+        return None
 
-        if (os.path.exists(greeter_path)):
-            # configure first found lightdm-greeter
-            for entry in os.listdir(greeter_path):
-                if entry.endswith('.desktop'):
-                    greeter = entry.split('.')[0]
-                    libcalamares.utils.debug(
-                        "found greeter {!s}".format(greeter)
-                        )
-                    os.system(
-                        "sed -i -e \"s/^.*greeter-session=.*"
-                        "/greeter-session={!s}/\" {!s}".format(
-                            greeter,
-                            lightdm_conf_path
-                            )
-                        )
-                    libcalamares.utils.debug(
-                        "{!s} configured as greeter.".format(greeter)
-                        )
-                    break
-            else:
-                return (
-                    _("Cannot configure LightDM"),
-                    _("No LightDM greeter installed.")
-                    )
 
+    def greeter_setup(self):
+        lightdm_conf_path = os.path.join(self.root_mount_point, "etc/lightdm/lightdm.conf")
+        greeter_path = self.find_preferred_greeter()
+
+        if greeter_path is not None and os.path.exists(greeter_path):
+            greeter = os.path.basename(os.path.realpath(greeter_path)) # Follow symlinks, hope they are not absolute
+            if greeter.endswith('.desktop'):
+                greeter = greeter[:-8] # Remove ".desktop" from end
+
+            libcalamares.utils.debug("found greeter {!s}".format(greeter))
+            os.system(
+                "sed -i -e \"s/^.*greeter-session=.*"
+                "/greeter-session={!s}/\" {!s}".format(
+                    greeter,
+                    lightdm_conf_path
+                )
```

Comment:
This should probably be done in python instead of using `sed` but at the very least it should be calling a libcalamares function instead of `os.system`.

Reply:
We don't really have a sed-like thing, do we (maybe we can link to a libsed or some suitable line-by-line replacer?). This module and the tracking module (! in C++) are the primary users of sed, so that would be a good improvement (also, eventually, towards dropping root).

But I'm not going to fix it in this PR, that's a larger issue.